pkg/runtime.v2/core/trainingruntime_test.go

```diff
@@ -53,59 +67,246 @@ func TestTrainingRuntimeNewObjects(t *testing.T) {
 				SpecAnnotation("conflictAnnotation", "override").
 				Trainer(
 					testingutil.MakeTrainJobTrainerWrapper().
-						ContainerImage("test:trainjob").
+						NumNodes(30).
 						Obj(),
 				).
 				Obj(),
-			trainingRuntime: baseRuntime.
-				Label("conflictLabel", "overridden").
-				Annotation("conflictAnnotation", "overridden").
-				RuntimeSpec(
-					testingutil.MakeTrainingRuntimeSpecWrapper(baseRuntime.Spec).
-						ContainerImage("test:runtime").
-						PodGroupPolicyCoschedulingSchedulingTimeout(120).
-						MLPolicyNumNodes(20).
-						ResourceRequests(0, corev1.ResourceList{
-							corev1.ResourceCPU: resource.MustParse("1"),
-						}).
-						ResourceRequests(1, corev1.ResourceList{
-							corev1.ResourceCPU: resource.MustParse("2"),
-						}).
-						Obj(),
-				).Obj(),
 			wantObjs: []client.Object{
 				testingutil.MakeJobSetWrapper(metav1.NamespaceDefault, "test-job").
+					NumNodes(30).
+					ContainerTrainer("test:runtime", []string{"runtime"}, []string{"runtime"}, resRequests).
+					ContainerDatasetModelInitializer("test:runtime", []string{"runtime"}, []string{"runtime"}, resRequests).
 					Suspend(true).
 					Label("conflictLabel", "override").
 					Annotation("conflictAnnotation", "override").
 					PodLabel(schedulerpluginsv1alpha1.PodGroupLabel, "test-job").
-					ContainerImage(ptr.To("test:trainjob")).
-					JobCompletionMode(batchv1.IndexedCompletion).
-					ResourceRequests(0, corev1.ResourceList{
-						corev1.ResourceCPU: resource.MustParse("1"),
-					}).
-					ResourceRequests(1, corev1.ResourceList{
-						corev1.ResourceCPU: resource.MustParse("2"),
-					}).
-					ControllerReference(kubeflowv2.SchemeGroupVersion.WithKind("TrainJob"), "test-job", "uid").
+					ControllerReference(kubeflowv2.SchemeGroupVersion.WithKind(kubeflowv2.TrainJobKind), "test-job", "uid").
 					Obj(),
 				testingutil.MakeSchedulerPluginsPodGroup(metav1.NamespaceDefault, "test-job").
-					ControllerReference(kubeflowv2.SchemeGroupVersion.WithKind("TrainJob"), "test-job", "uid").
-					MinMember(40).
-					SchedulingTimeout(120).
+					ControllerReference(kubeflowv2.SchemeGroupVersion.WithKind(kubeflowv2.TrainJobKind), "test-job", "uid").
+					MinMember(31). // 31 replicas = 30 Trainer nodes + 1 Initializer.
 					MinResources(corev1.ResourceList{
-						corev1.ResourceCPU: resource.MustParse("60"),
+						corev1.ResourceCPU: resource.MustParse("31"), // Every replica has 1 CPU = 31 CPUs in total.
```

Comment:
I think we should have a utility function to calculate the quantity passing in the obj

Reply:
We already have it here to calculate PodRequests based on the PodSpec: https://github.com/kubeflow/training-operator/blob/master/pkg/runtime.v2/runtime.go#L121-L124.

Do you mean having another utility just for the unit tests ?

Reply:
I was not sure the reason why @akshaychitneni requests to implement the util.

Reply:
@akshaychitneni Please can you give an example of code snippet that you want to use instead of:
```
corev1.ResourceCPU: resource.MustParse("31")
```

Reply:
I was referring to calculating 31. Probably it is simpler as it is `(numNodes + 1) * (Num of CPU configured in the runtime) ` in this case

Reply:
I see, I think it depends on the unit test. For example, we have unit tests where we add CPU + Memory for the Pod requests: https://github.com/kubeflow/training-operator/blob/85d0cf1fca0e56d206e579f77d68f33aca1a0f87/pkg/runtime.v2/framework/core/framework_test.go#L359-L360.
Additionally, in the future we might want to define different resources for initializer containers and trainer to verify that we calculate MinResource correctly for PodGroup.

Let me add the TODO statement to see if we can create helper function for our tests to calculate these values based on number of replicas and resources per replica.

Reply:
Actually, container resource computation is not easy, we need to consider runtimeClass (not training runtime, it's core API runtimeClass), limitRange, and initContainer and so on. So, I would like to keep manually resource creation for the testing.

If we provide such utils with complex computing, the lib could potentially bug, and we will miss the bugs in main testing cases.